File: Cpp_Algo/백준/Silver/5014. 스타트링크/스타트링크.py
from collections import deque

# 최대 F층 / 현재 위치 S층 / 목적지 G층 / 위로 U층 이동 가능 / 아래로 D층 이동 가능
F, S, G, U, D = map(int, input().split())
visited = [1000001] * (F + 1)
visited[S] = 0

queue = deque()
queue.append(S)

while queue:
    now = queue.popleft()

    if (now + U) < (F + 1) and visited[now + U] == 1000001:
        queue.append(now + U)
        visited[now + U] = visited[now] + 1

    if (now - D) > 0 and visited[now - D] == 1000001:
        queue.append(now - D)
        visited[now - D] = visited[now] + 1

if visited[G] == 1000001:
    print("use the stairs")
else:
    print(visited[G])

File: 백준/Silver/5014. 스타트링크/스타트링크.cc
#include <iostream>
using namespace std;

const int MAX = 1000001;

int F, S, G, U, D;
int visited[MAX];
int queue[MAX];

int main() {
    cin.tie(0);
    cout.tie(0);
    ios::sync_with_stdio(false);

    int front = 0;
    int rear = 0;

    int now;

    cin >> F >> S >> G >> U >> D;

    for (int i = 0; i < MAX; i++)
    {
        visited[i] = MAX;
    }

    visited[S] = 0;
    queue[rear++] = S;
    
    while (front != rear) {
        now = queue[front++];

        if (now + U < F + 1 && visited[now + U] == MAX) {
            queue[rear++] = now + U;
            visited[now + U] = visited[now] + 1;
        }

        if (now - D > 0 && visited[now - D] == MAX) {
            queue[rear++] = now - D;
            visited[now - D] = visited[now] + 1;
        }
    }

    if (visited[G] == MAX) {
        cout << "use the stairs";
    }
    else {
        cout << visited[G];
    }

    return 0;
}
